lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+  // Verify patch constant signature
+  VerifySignature(DM.GetPatchConstOrPrimSignature(),
+                  PSV.GetPatchConstOrPrimElement0(0),
+                  PSV.GetSigPatchConstOrPrimElements(),
+                  "SigPatchConstantOrPrim", i1ToUnknownCompat);
+}
+
+void PSVContentVerifier::VerifySignature(const DxilSignature &Sig,
+                                         PSVSignatureElement0 *Base,
+                                         unsigned Count, std::string Name,
+                                         bool i1ToUnknownCompat) {
+  if (Count != Sig.GetElements().size()) {
+    EmitMismatchError(Name + "Elements", std::to_string(Count),
+                      std::to_string(Sig.GetElements().size()));
+    return;
+  }
+
+  // Verify each element in DxilSignature.
+  const PSVStringTable &StrTab = PSV.GetStringTable();
+  const PSVSemanticIndexTable &IndexTab = PSV.GetSemanticIndexTable();
+  for (unsigned i = 0; i < Count; i++)
+    VerifySignatureElement(Sig.GetElement(i),
+                           PSV.GetRecord<PSVSignatureElement0>(
+                               Base, PSV.GetSignatureElementSize(), Count, i),
+                           StrTab, IndexTab, Name, i1ToUnknownCompat);
+}
+
+void PSVContentVerifier::VerifySignatureElement(
+    const DxilSignatureElement &SE, PSVSignatureElement0 *PSVSE0,
+    const PSVStringTable &StrTab, const PSVSemanticIndexTable &IndexTab,
+    std::string Name, bool i1ToUnknownCompat) {
+  // Find the signature element in the set.
+  PSVSignatureElement0 ModulePSVSE0;
+  InitPSVSignatureElement(ModulePSVSE0, SE, i1ToUnknownCompat);
+
+  // Check the Name and SemanticIndex.
+  bool Mismatch = false;
+  const std::vector<uint32_t> &SemanticIndexVec = SE.GetSemanticIndexVec();
+  llvm::ArrayRef<uint32_t> PSVSemanticIndexVec(
+      IndexTab.Get(PSVSE0->SemanticIndexes), PSVSE0->Rows);
+  if (SemanticIndexVec.size() == PSVSemanticIndexVec.size())
+    Mismatch |= memcmp(PSVSemanticIndexVec.data(), SemanticIndexVec.data(),
+                       SemanticIndexVec.size() * sizeof(uint32_t)) != 0;
+  else
+    Mismatch = true;
+
+  ModulePSVSE0.SemanticIndexes = PSVSE0->SemanticIndexes;
+
+  PSVSignatureElement PSVSE(StrTab, IndexTab, PSVSE0);
+  if (SE.IsArbitrary())
+    Mismatch |= strcmp(PSVSE.GetSemanticName(), SE.GetName());
+  else
+    Mismatch |= PSVSE0->SemanticKind != static_cast<uint8_t>(SE.GetKind());
+
+  ModulePSVSE0.SemanticName = PSVSE0->SemanticName;
+  // Compare all fields.
+  Mismatch |= memcmp(&ModulePSVSE0, PSVSE0, sizeof(PSVSignatureElement0)) != 0;
+  if (Mismatch) {
+    PSVSignatureElement ModulePSVSE(StrTab, IndexTab, &ModulePSVSE0);
+    std::string ModuleStr;
+    raw_string_ostream OS(ModuleStr);
+    ModulePSVSE.Print(OS, SE.GetName(), SemanticIndexVec.data());
+    OS.flush();
+    std::string PartStr;
+    raw_string_ostream OS1(PartStr);
+    PSVSE.Print(OS1, PSVSE.GetSemanticName(), PSVSemanticIndexVec.data());
```

Comment:
I see the introduction of this deconstructed print was added so that there could be just one error printing condition regardless of whether it was prompted by a difference in semantic indexes, semantic names, or something else. I think that's a lot of churn to prevent having custom fail cases above for the semantic index and name failure cases, dumping the entire Signature Element would be pretty confusing while we have more specific information.